src/main/permissionsManager.ts

```diff
@@ -0,0 +1,170 @@
+
+        // allow if the request is coming from the local renderer process instead of the remote one
+        const mainWindow = MainWindow.get();
+        if (mainWindow && webContentsId === mainWindow.webContents.id) {
+            return true;
+        }
+
+        const parsedURL = parseURL(requestingURL);
+        if (!parsedURL) {
+            return false;
+        }
+
+        let serverURL;
+        if (CallsWidgetWindow.isCallsWidget(webContentsId)) {
+            serverURL = CallsWidgetWindow.getViewURL();
+        } else {
+            serverURL = ViewManager.getViewByWebContentsId(webContentsId)?.view.server.url;
+        }
+
+        if (!serverURL) {
+            return false;
+        }
+
+        // is the requesting url trusted?
+        if (!isTrustedURL(parsedURL, serverURL)) {
+            return false;
+        }
+
+        // For certain permission types, we need to confirm with the user
+        if (authorizablePermissionTypes.includes(permission)) {
+            const currentPermission = this.json[parsedURL.origin]?.[permission];
+
+            // If previously allowed, just allow
+            if (currentPermission?.allowed) {
+                return true;
+            }
+
+            // If denied permanently, deny
+            if (currentPermission?.alwaysDeny) {
+                return false;
+            }
+
+            if (!mainWindow) {
+                return false;
+            }
+
+            // Show the dialog to ask the user
+            const {response} = await dialog.showMessageBox(mainWindow, {
+                title: localizeMessage('main.permissionsManager.checkPermission.dialog.title', 'Permission Requested'),
+                message: localizeMessage(`main.permissionsManager.checkPermission.dialog.message.${permission}`, '{appName} ({url}) is requesting the "{permission}" permission.', {appName: app.name, url: parsedURL.origin, permission}),
+                detail: localizeMessage(`main.permissionsManager.checkPermission.dialog.detail.${permission}`, 'Would you like to grant {appName} this permission?', {appName: app.name}),
+                type: 'question',
+                buttons: [
+                    localizeMessage('label.allow', 'Allow'),
+                    localizeMessage('label.deny', 'Deny'),
+                    localizeMessage('label.denyPermanently', 'Deny Permanently'),
+                ],
+            });
+
+            // Save their response
+            const newPermission = {
+                allowed: response === 0,
+                alwaysDeny: (response === 2) ? true : undefined,
+            };
+            this.json[parsedURL.origin] = {
+                ...this.json[parsedURL.origin],
+                [permission]: newPermission,
+            };
+            this.writeToFile();
+
+            if (response > 0) {
+                return false;
+            }
+        }
+
+        // We've checked everything so we're okay to grant the remaining cases
+        return true;
```

Comment:
Which are the remaining cases?

Reply:
Haven't we already returned `true` for all these 3 cases?

Reply:
The only two we return true for earlier are:
- if it's the main window
- if the permission was previously allowed by the user

Reply:
Sounds good, thanks for explaining